### PR TITLE
Adding benchmarking for new `frame_election_provider_support`

### LINKED ISSUE #11111: Benchmark different implementations of `Solver`
All structs that implement `NposSolver` are currently returning a naive `max_block` weight, just because they don't have any better way to estimate them. 

The weight of these implementations are not really runtime dependent, we can and should still leverage benchmarking pipeline. So, I propose the follwoing: 

2. Create a sub-create in `election-provider-support` called `onchain-config`. This will have a `trait WeightInfo`, similar to all other pallets, with 2 functions

```
trait WeightInfo {
  fn phragmen(<input>) -> Weight;
  fn phragmms(<input>) -> Weight;
}
```

4. This is a raw pallet, with nothing in it other than a:

```
trait Config: frame_system::Config { 
  type WeightInfo: WeightInfo;
  // ++ everything else that's in `trait Config` in `onchain.rs` now.
  /// The accuracy used to compute the election:
  type Accuracy: PerThing128;
  /// Something that provides the data for election.
  type DataProvider: ElectionDataProvider<
    AccountId = Self::AccountId,
    BlockNumber = Self::BlockNumber,
  >;
}
```

6. Then you write benchmarks for this pallet, just as you would do for any other pallet, trying to fulfill the aforementioned `trait WeightInfo`. If any further information is needed to conduct the benchmarks accurately, you can pass in more types to `trait Config`.
7. As you can probably guess now, the existing `trait Config` in `onchain.rs` can be removed. Any struct wanting to be a `OnchainExecution` needs to implement to new `onchain_config::Config` pallet config. 

The main thing that I didn't cover is what is `<input>` in the above snippet. These should be the components that dictate the complexity of phragmen, or similar algorithms. I have tried to do this once in the past, and my conclusion was that this is a function of 3 things: 

1. number of voters
2. number of candidates
3. number of edges (votes per voter)

A similar benchmark was done here: https://github.com/paritytech/substrate/blob/master/frame/elections-phragmen/src/benchmarking.rs#L454

For more general info about benchmarking, see: https://docs.substrate.io/v3/runtime/benchmarking/, https://www.shawntabrizi.com/substrate-graph-benchmarks/docs/#/

### PR DESCRIPTION
**Pull request by georgesdib**
Fixes #11111

Created a benchmarking pallet for `frame-election-provider-support`, namely `SequentialPhragmen` and `PhragMMS`.
Few changes to `onchain.rs` as well and making the trait always bounded, and adding `WeightInfo`.

Refactoring #11126 

##

Polkadot companion: paritytech/polkadot#5241
skip check-dependent-cumulus

Polkadot address: 131dPecTmpTC1p1ofemufqFBJo9vNbV7dkgN7vWwKnaSMkC4

### CODE REVIEW

**Comment by kianenigma** [comment]
I have some minor feedback as well, but I am going to save that until this major flaw is solved: the purpose of this pallet is to benchmark these election algorithms alone and isolated. The reason for this is that currently the `ElectionDataProvider` trait specifies that its implementors need to be self-weighing and register their own weight. With that out of the way, we need to only benchmark `T::Solver::solve()` to have some kind of a realistic measure.

**Comment by kianenigma** [comment]
Here's some random feedback and how I think this should work out, but this does not compile yet and has some TODOs. If you agree with the overall direction, the branch can be a good guideline: 

https://github.com/georgesdib/substrate/pull/new/kiz-onchain-election-benchmarking


Lastly: 

> For now, we have `VotersBound` and `TargetsBound` as being part of ConfigParams which is normally implemented for Runtime. However, ideally these 2 types should be part of the `ElectionDataProvider` trait as they are really a property of the data provider. This could be worked on in a separate PR with a big ensuing refactoring.

A `VoterBound` on `ElectionDataProvider` has the meaning of: what is the maximum number of voters that this data provider can provide (which I don't think we ever need to bound)

A `VoterBound` on `onchain::Config` has the meaning of: how many voters should this implementation of `onchain::Config` ask from its `ElectionDataProvider`? 

TLDR; I think `VotersBound` on `onchain::Config` makes good sense :)

File: bin/node/runtime/src/lib.rs [hunk 1]
```diff
@@ -1847,6 +1846,7 @@ impl_runtime_apis! {
 			// which is why we need these two lines below.
 			use pallet_session_benchmarking::Pallet as SessionBench;
 			use pallet_offences_benchmarking::Pallet as OffencesBench;
+			use pallet_election_provider_support_benchmarking::Pallet as PepsoBench;
```
Comment (kianenigma) [inline bin/node/runtime/src/lib.rs:1849]: we can drop the the `P`, signifying `Pallet`, and call this `EPSBench`: Election provider support.

File: frame/election-provider-multi-phase/src/mock.rs [hunk 2]
```diff
@@ -276,12 +276,17 @@ parameter_types! {
...
+impl onchain::Config for OnChainSeqPhragmen {
 	type System = Runtime;
 	type Solver = SequentialPhragmen<AccountId, SolutionAccuracyOf<Runtime>, Balancing>;
 	type DataProvider = StakingMock;
+	type VotersBound = ConstU32<600>;
+	type TargetsBound = ConstU32<100_000>;
+	type WeightInfo = ();
 }
 
+type OnChainPhragmen = onchain::BoundedExecution<OnChainSeqPhragmen>;
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/mock.rs:288]: same

File: frame/election-provider-support/src/onchain.rs [hunk 3]
```diff
@@ -41,39 +45,19 @@ impl From<sp_npos_elections::Error> for Error {
...
-/// Configuration trait of [`UnboundedExecution`].
-pub trait ExecutionConfig {
-	/// Something that implements the system pallet configs. This is to enable to register extra
-	/// weight.
+/// dynamically. If you use `elect_with_bounds` along with `InstantElectionProvider`, the bound that
+/// would be used is the minimum of the dynamic bounds given as arguments to `elect_with_bounds` and
+/// the trait bounds (`MaxVoters` and `MaxTargets`).
+pub struct BoundedExecution<T: Config>(PhantomData<T>);
+
+/// Configuration trait.
```
Comment (kianenigma) [inline frame/election-provider-support/src/onchain.rs:58]:
```suggestion
/// Configuration trait for an onchain election execution.
```

File: bin/node/runtime/src/lib.rs [hunk 4]
```diff
@@ -557,7 +557,7 @@ impl pallet_staking::Config for Runtime {
 	type MaxNominatorRewardedPerValidator = MaxNominatorRewardedPerValidator;
 	type OffendingValidatorsThreshold = OffendingValidatorsThreshold;
 	type ElectionProvider = ElectionProviderMultiPhase;
-	type GenesisElectionProvider = onchain::UnboundedExecution<OnChainSeqPhragmen>;
+	type GenesisElectionProvider = onchain::BoundedExecution<OnChainSeqPhragmen>;
```
Comment (kianenigma) [inline bin/node/runtime/src/lib.rs:560]: why?

File: frame/grandpa/src/mock.rs [hunk 5]
```diff
@@ -181,12 +181,17 @@ parameter_types! {
...
 	type System = Test;
-	type Solver = SequentialPhragmen<<Test as frame_system::Config>::AccountId, Perbill>;
+	type Solver = SequentialPhragmen<u64, Perbill>;
 	type DataProvider = Staking;
+	type VotersBound = ConstU32<600>;
+	type TargetsBound = ConstU32<400>;
+	type WeightInfo = ();
 }
 
+type OnChainPhragmen = onchain::BoundedExecution<OnChainSeqPhragmen>;
```
Comment (kianenigma) [inline frame/grandpa/src/mock.rs:193]: same.

File: frame/offences/benchmarking/src/mock.rs [hunk 6]
```diff
@@ -151,12 +151,17 @@ parameter_types! {
 pub type Extrinsic = sp_runtime::testing::TestXt<Call, ()>;
 
 pub struct OnChainSeqPhragmen;
-impl onchain::ExecutionConfig for OnChainSeqPhragmen {
+impl onchain::Config for OnChainSeqPhragmen {
 	type System = Test;
 	type Solver = SequentialPhragmen<AccountId, Perbill>;
 	type DataProvider = Staking;
+	type VotersBound = ConstU32<600>;
```
Comment (kianenigma) [inline frame/offences/benchmarking/src/mock.rs:158]: These should be `u32::MAX` IMO instead of an arbitrary small number.

File: frame/offences/benchmarking/src/mock.rs [hunk 7]
```diff
@@ -151,12 +151,17 @@ parameter_types! {
...
+	type TargetsBound = ConstU32<400>;
+	type WeightInfo = ();
 }
 
+type OnChainPhragmen = onchain::BoundedExecution<OnChainSeqPhragmen>;
```
Comment (kianenigma) [inline frame/offences/benchmarking/src/mock.rs:163]: ditto.

File: frame/election-provider-support/src/onchain.rs [hunk 8]
```diff
@@ -41,39 +45,19 @@ impl From<sp_npos_elections::Error> for Error {
...
-pub struct BoundedExecution<T: BoundedExecutionConfig>(PhantomData<T>);
-
-/// An unbounded variant of [`BoundedExecution`].
-///
-/// ### Warning
-///
-/// This can be very expensive to run frequently on-chain. Use with care. Moreover, this
-/// implementation ignores the additional data of the election data provider and gives no insight on
-/// how much weight was consumed.
-pub struct UnboundedExecution<T: ExecutionConfig>(PhantomData<T>);
```
Comment (kianenigma) [inline frame/election-provider-support/src/onchain.rs:71]: not sure, and not a fan of this going away, why?

**Review by kianenigma — COMMENTED** [review]
Last round of comments, mostly looks good!

File: frame/election-provider-support/benchmarking/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+#![cfg(feature = "runtime-benchmarks")]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::Decode;
+use frame_benchmarking::{benchmarks, Vec};
+use frame_election_provider_support::{NposSolver, PhragMMS, SequentialPhragmen};
+
+pub struct Pallet<T: Config>(frame_system::Pallet<T>);
+pub trait Config: frame_system::Config {}
```
[the remark this answers is not in the document]
Comment (georgesdib) [reply frame/election-provider-support/benchmarking/src/lib.rs:29]: The issues are:
- The benchmark system seems to impose that the `Config` trait is implemented for `Runtime`, however the `onchain::Config` is not (and by design because otherwise we cannot have 2 different configs for `Runtime` for example the 2 algorithms).
- Don't we ideally want to have the benchmarking bounds different from the trait bounds? For example say the pallet is unbounded (bounded by the max value), we would still want to cap the benchmarking to not take a lot of time to run.
- I thought about adding `BenchmarkingConfig` to configure that, but then we need `BenchmarkingConfig` to be implemented in the node runtime because each chain may want to benchmark it differently, and we need the benchmarking pallet to have access to it, so it probably needs to be implemented for `Runtime`, this looked like over engineering it for now, and having these values defined as constant in the pallet itself acceptable.

Obviously happy to take suggestions to the contrary.

File: frame/election-provider-support/src/onchain.rs (hunk 8 above)
Comment (georgesdib) [reply frame/election-provider-support/src/onchain.rs:71]: A bit of an opinionated commit indeed, does it make sense to keep an unbounded onchain execution? The bounds could always be chosen to be arbitrarily large but it enforces some "discipline" on the part of the pallet user.
More than happy to revert this change though!

File: frame/election-provider-support/benchmarking/src/lib.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (kianenigma) [reply frame/election-provider-support/benchmarking/src/lib.rs:29]: Okay, I see the problem. For now let's keep it as-is, and keep the benchmarking configs "constant", and later on we can add them to this `trait Config` if needed.

**Review by kianenigma — APPROVED** [review]

**Comment by kianenigma** [comment]
\tip large

**Comment by kianenigma** [comment]
\tip large

File: frame/election-provider-support/benchmarking/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+	"sp-npos-elections/std",
+	"sp-runtime/std",
+	"frame-benchmarking/std",
+	"frame-system/std",
+]
+
+runtime-benchmarks = [
+	"frame-benchmarking/runtime-benchmarks",
+    "frame-election-provider-support/runtime-benchmarks",
+]
```
Comment (emostov) [inline frame/election-provider-support/benchmarking/Cargo.toml:37]:
```suggestion
]

```

File: frame/election-provider-support/benchmarking/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+		})
+		.collect::<Vec<_>>();
+
+	(voters, targets)
+}
+
+benchmarks! {
+	phragmen {
+		// number of votes in snapshot.
+		let v in (VOTERS[0]) .. VOTERS[1];
```
Comment (emostov) [inline frame/election-provider-support/benchmarking/src/lib.rs:62]:
```suggestion
		let v in (VOTERS[0]) .. VOTERS[1];
```
I assume the `(` around `(VOTERS[0]) ` are just to get the macro to compile?

Comment (georgesdib) [reply frame/election-provider-support/benchmarking/src/lib.rs:62]: Indeed

**Review by emostov — APPROVED** [review]
LGTM - thanks

File: frame/election-provider-support/src/onchain.rs (hunk 8 above)
Comment (georgesdib) [reply frame/election-provider-support/src/onchain.rs:71]: I have reverted back `UnboundedConfig`.

**Comment by kianenigma** [comment]
\tip large

**Comment by kianenigma** [comment]
Okay, I will open a manual tip for you @georgesdib.

**Comment by kianenigma** [comment]
bot merge

**Comment by kianenigma** [comment]
bot merge

**Comment by kianenigma** [comment]
bot merge

**Comment by kianenigma** [comment]
bot merge

**Comment by kianenigma** [comment]
bot merge

**Comment by ggwpez** [comment]
You skipped the cumulus check, somehow my [cumulus CI](https://github.com/paritytech/cumulus/pull/1172) errors with:  
```pre
error[E0432]: unresolved import `frame_election_provider_support::onchain::ExecutionConfig`
  --> /usr/local/cargo/git/checkouts/polkadot-4038f27d5e4ea2e8/3ff2b83/runtime/common/src/elections.rs:20:12
   |
20 |     onchain::{ExecutionConfig, UnboundedExecution},
   |               ^^^^^^^^^^^^^^^ no `ExecutionConfig` in `onchain`
```
Could this be related @georgesdib ?

**Comment by georgesdib** [comment]
> You skipped the cumulus check, somehow my [cumulus CI](https://github.com/paritytech/cumulus/pull/1172) errors with:
> 
> ```
> error[E0432]: unresolved import `frame_election_provider_support::onchain::ExecutionConfig`
>   --> /usr/local/cargo/git/checkouts/polkadot-4038f27d5e4ea2e8/3ff2b83/runtime/common/src/elections.rs:20:12
>    |
> 20 |     onchain::{ExecutionConfig, UnboundedExecution},
>    |               ^^^^^^^^^^^^^^^ no `ExecutionConfig` in `onchain`
> ```
> 
> Could this be related @georgesdib ?

Isn't this in from the polkadot codebase? The companion PR is now merged, perhaps the above was run after the substrate PR was merged but not the polkadot PR?

**Comment by ggwpez** [comment]
Yes sorry, I think the dependencies just got messed up. Thanks for you contributions!
